extensions/amp-form/0.1/form-validators.js

```diff
@@ -0,0 +1,370 @@
+ * @interface
+ */
+class FormValidator {
+
+  /**
+   * @param {!HTMLFormElement} form
+   */
+  constructor(form) {
+    /** @protected @const {!HTMLFormElement} */
+    this.form = form;
+
+    /** @protected @const {!Document} */
+    this.doc = form.ownerDocument;
+  }
+
+  /**
+   * Called to report validation errors.
+   */
+  report() {}
+
+  /**
+   * @param {!Event} unusedEvent
+   */
+  onBlur(unusedEvent) {}
+
+  /**
+   * @param {!Event} unusedEvent
+   */
+  onInput(unusedEvent) {}
+}
+
+
+class DefaultValidator extends FormValidator {
+
+  /** @override */
+  report() {
+    this.form.reportValidity();
+  }
+
+}
+
+
+class PolyfillDefaultValidator extends FormValidator {
+
+  constructor(form) {
+    super(form);
+
+    const win = this.doc.defaultView;
+    const bubbleId = `amp-validation-bubble-${validationBubbleCount++}`;
+    /** @private @const {!./validation-bubble.ValidationBubble} */
+    this.validationBubble_ = new ValidationBubble(win, bubbleId);
+  }
+
+  /** @override */
+  report() {
+    const inputs = this.form.querySelectorAll('input,select,textarea');
+    for (let i = 0; i < inputs.length; i++) {
+      if (!inputs[i].checkValidity()) {
+        inputs[i]./*REVIEW*/focus();
+        this.validationBubble_.show(inputs[i], inputs[i].validationMessage);
+        break;
+      }
+    }
+  }
+
+  /** @override */
+  onBlur(unusedEvent) {
+    this.validationBubble_.hide();
+  }
+
+  /** @override */
+  onInput(event) {
+    if (!this.validationBubble_.isActiveOn(event.target)) {
+      return;
+    }
+
+    if (event.target.checkValidity()) {
+      event.target.removeAttribute('aria-invalid');
+      this.validationBubble_.hide();
+    } else {
+      event.target.setAttribute('aria-invalid', 'true');
+      this.validationBubble_.show(event.target, event.target.validationMessage);
+    }
+  }
+}
+
+
+/**
+ * @abstract
+ */
+class AbstractCustomValidator extends FormValidator {
+
+  constructor(form) {
+    super(form);
+
+    /** @private @const {!Object<string, !Element>} */
+    this.inputValidationsDict_ = {};
+
+    /** @private @const {!Array<!Element>} */
+    this.inputVisibleValidationDict_ = {};
+  }
+
+  /**
+   * @param {!Element} input
+   */
+  reportInput(input) {
+    const invalidType = getInvalidType(input);
+    if (invalidType) {
+      this.showValidationFor(input, invalidType);
+    }
+  }
+
+  /**
+   * Hides all validation messages.
+   */
+  hideAllValidations() {
+    const inputs = this.form.querySelectorAll('input,select,textarea');
+    for (let i = 0; i < inputs.length; i++) {
+      this.hideValidationFor(inputs[i]);
+    }
+  }
+
+  /**
+   * @param {!Element} input
+   * @param {!string} invalidType
+   * @return {?Element}
+   */
+  getValidationFor(input, invalidType) {
+    if (!input.id) {
+      return null;
+    }
+    const selector = `[visible-when-invalid=${invalidType}]` +
+        `[validation-for=${input.id}]`;
+    if (!this.inputValidationsDict_[selector]) {
```

Comment:
This should be testing for `=== undefined`, so that we don't repeatedly issue queries for something that doesn't have a `[visibile-when-invalid=X]`